### PR TITLE
fix(compiler): treat i18n attributes with no bindings as static attributes

### LINKED ISSUE #38231: i18n with ivy breaks the injection of translated attribute
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/localize.

### Is this a regression?

Yes, this issue is reproducible only with Ivy. With ViewEngine it works fine.

### Description

Injecting the attribute to the component/directive that is translated using `i18n` is always resulting in `null`.

Example:

`text` is a directive that reads the `text` attribute and simply adds the following text to the host element:

> Text attribtue value is: <attribute value>

Code:

```html
<div text="Text without i18n"></div>
<div i18n-text text="Text with i18n"></div>
```

Resulting HTML:

```html
<div text="Text without i18n">Text attribtue value is: Text without i18n.</div>
<div text="Text with i18n">Text attribtue value is: null.</div>
```

Observed in the result above, `text` attribute is properly set to `Text with i18n`, while the injected value remains `null`.

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-ivy-uzgvwe

You can observe `null` as described above.

To make this example work, open settings and uncheck `Enable Ivy`. Now ViewEngine will be used to compile the application, and the `null` text will be replaced by `Text with i18n`.

## 🌍  Your Environment

Angular CLI: 10.0.4
Node: 12.16.3
OS: linux x64

Angular: 10.0.5
... animations, common, compiler, compiler-cli, core, forms
... localize, platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1000.4
@angular-devkit/build-angular     0.1000.4
@angular-devkit/build-optimizer   0.1000.4
@angular-devkit/build-webpack     0.1000.4
@angular-devkit/core              10.0.4
@angular-devkit/schematics        10.0.4
@angular/cli                      10.0.4
@ngtools/webpack                  10.0.4
@schematics/angular               10.0.4
@schematics/update                0.1000.4
rxjs                              6.5.5
typescript                        3.9.7
webpack                           4.43.0

### PR DESCRIPTION
**Pull request by crisbeto**
Currently `i18n` attributes are treated the same no matter if they have data bindings or not. This both generates more code since they have to go through the `ɵɵi18nAttributes` instruction and prevents the translated attributes from being injected using the `@Attribute` decorator.

These changes makes it so that static translated attributes are treated in the same way as regular static attributes and all other `i18n` attributes go through the old code path.

Fixes #38231.

### CODE REVIEW

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 1]
```diff
@@ -153,15 +153,22 @@ ${output}
...
 };
 
-const verify = (input: string, output: string, extra: any = {}): void => {
+const verify = (input: string, output: string, extra: {
+  inputArgs?: any,
+  compilerOptions?: any,
+  skipPathBasedCheck?: boolean,
+  skipIdBasedCheck?: boolean,
+  verbose?: boolean,
+  exceptions?: {}
```
Comment (crisbeto) [inline packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:162]: These changes aren't necessary for the fix, I just decided to add better typing since I found myself struggling to find the options when I was debugging.

File: packages/core/test/acceptance/i18n_spec.ts [hunk 2]
```diff
@@ -1843,40 +1843,6 @@ onlyInIvy('Ivy i18n logic').describe('runtime i18n', () => {
       expect(fixture.nativeElement.firstChild.title).toEqual(`ANGULAR - value 1 - value 2 (fr)`);
     });
 
-    it('should create corresponding ng-reflect properties', () => {
```
Comment (crisbeto) [inline packages/core/test/acceptance/i18n_spec.ts:1846]: This test started failing, because static attributes don't generate `ng-reflect` attributes. I think that it makes sense to remove it, given that the attributes are treated as static now.

File: packages/compiler/src/render3/view/template.ts [hunk 3]
```diff
@@ -866,10 +864,18 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
     this.matchDirectives(NG_TEMPLATE_TAG_NAME, template);
 
     // prepare attributes parameter (including attributes used for directive matching)
-    const [i18nStaticAttrs, staticAttrs] = partitionArray(template.attributes, hasI18nMeta);
+    const boundI18nAttrs: t.BoundAttribute[] = [];
```
Comment (AndrewKushnir) [inline packages/compiler/src/render3/view/template.ts:867]: Can we still use `partitionArray`, but just a different condition function?

File: packages/compiler/src/render3/view/template.ts [hunk 4]
```diff
@@ -1289,18 +1295,23 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
...
+    for (const attr of renderAttributes) {
       if (attr.name === NG_PROJECT_AS_ATTR_NAME) {
         ngProjectAsAttr = attr;
       }
-      attrExprs.push(
-          ...getAttributeNameLiterals(attr.name), trustedConstAttribute(elementName, attr));
-    });
+
+      if (attr.i18n) {
+        attrExprs.push(o.literal(attr.name), this.i18nTranslate(attr.i18n as i18n.Message));
```
Comment (AndrewKushnir) [inline packages/compiler/src/render3/view/template.ts:1309]: I'd propose to add a quick comment here to explain why i18n attrs are not in the `i18n` section of the attrs array. It's clear in a context of this PR, but might be tricky when we look at the code later :)

File: packages/core/test/acceptance/i18n_spec.ts [hunk 5]
```diff
@@ -3024,6 +2990,27 @@ onlyInIvy('Ivy i18n logic').describe('runtime i18n', () => {
     fixture.detectChanges();
     expect(fixture.nativeElement.textContent).toEqual(`two,one,`);
   });
+
+  it('should be able to inject a static i18n attribute', () => {
+    loadTranslations({[computeMsgId('text')]: 'translatedText'});
+
+    @Directive({selector: '[injectTitle]'})
+    class InjectTitleDir {
+      constructor(@Attribute('title') public title: string) {}
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/i18n_spec.ts:2999]: I think we can add a couple more tests (or extend this test) for completeness:
- where the `@Attribute` is used to inject an i18n attr with interpolation (so we'll have `null` I guess)
- similar to the previous one, but a non-i18n attr with interpolation (to verify that we get the same result)

**Review by AndrewKushnir — COMMENTED** [review]
Looks great, thanks for this PR @crisbeto 👍 

I think we can also simplify the `i18nAttributes` instruction by removing the code that takes care of [the static value processing here](https://github.com/angular/angular/blob/master/packages/core/src/render3/i18n/i18n_parse.ts#L254-L270). There might be even more refactoring in that function that we can look at: we may not need the `hasBinding` check (since we know that it's a dynamic value, this we'd have bindings) and ICU-related logic there (we do not support ICUs in attrs atm).

File: packages/core/src/render3/i18n/i18n_parse.ts [hunk 6]
```diff
@@ -622,10 +599,10 @@ function walkIcuTree(
                   generateBindingUpdateOpCodes(update, attr.value, newIndex, attr.name);
                 }
               } else {
-                ngDevMode && console.warn(` WARNING:
-      ignoring unsafe attribute value ${lowerAttrName} on element $ {
-    tagName
-  } (see http://g.co/ng/security#xss)`);
+                ngDevMode &&
+                    console.warn(
```
Comment (crisbeto) [inline packages/core/src/render3/i18n/i18n_parse.ts:603]: Not a necessary change, but I noticed that the string wouldn't have generated the correct warning so I fixed it up.

File: packages/core/test/render3/i18n/i18n_spec.ts [hunk 7]
```diff
@@ -416,30 +416,6 @@ describe('Runtime i18n', () => {
   });
 
   describe(`i18nAttribute`, () => {
-    it('for text', () => {
```
Comment (crisbeto) [inline packages/core/test/render3/i18n/i18n_spec.ts:419]: I dropped this test, because it looked like it was testing code specific to static i18n attribute bindings. There's one right under it that is checking for i18n attributes with an interpolation.

**Review by crisbeto — COMMENTED** [review]
Thanks for the feedback @AndrewKushnir, all of it should be addressed now. Also something I was thinking about while addressing the feedback: would it make sense to rename the `i18nAttribute` instruction to something like `i18nDynamicAttribute` to be more explicit about its uses? It doesn't have to happen as a part of this PR.

File: packages/compiler/src/render3/view/i18n/util.ts [hunk 8]
```diff
@@ -50,6 +50,11 @@ export function hasI18nMeta(node: t.Node&{i18n?: i18n.I18nMeta}): boolean {
   return !!node.i18n;
 }
 
+export function isBoundI18nAttribute(node: t.TextAttribute|
+                                     t.BoundAttribute): node is t.BoundAttribute {
+  return node.i18n != null && !(node instanceof t.TextAttribute);
```
Comment (AndrewKushnir) [inline packages/compiler/src/render3/view/i18n/util.ts:55]: May be invert this condition and check for `t.BoundAttribute`?
```suggestion
  return node.i18n != null && (node instanceof t.BoundAttribute);
```

File: packages/core/src/render3/i18n/i18n_parse.ts [hunk 9]
```diff
@@ -244,30 +241,10 @@ export function i18nAttributesFirstPass(
...
-          const dataValue = tNode.inputs !== null && tNode.inputs[attrName];
-          if (dataValue) {
-            setInputsForProperty(tView, lView, dataValue, attrName, value);
-            if (ngDevMode) {
-              const element = getNativeByIndex(previousElementIndex, lView) as RElement | RComment;
-              setNgReflectProperties(lView, element, tNode.type, dataValue, value);
-            }
-          }
-        }
+      } else if (value !== '' && tView.firstCreatePass && tView.data[index] === null) {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n/i18n_parse.ts:244]: I think we can move this part of the check outside of the `for` loop:
```
const isFirstCreatePass = tView.firstCreatePass && tView.data[index] === null;

// ...

} else if (isFirstCreatePass && value !== '') {
// ...
```
and we can also reuse this in this function below.

File: packages/core/src/render3/i18n/i18n_parse.ts [hunk 10]
```diff
@@ -244,30 +241,10 @@ export function i18nAttributesFirstPass(
         // Odd indexes are ICU expressions
         // TODO(ocombe): support ICU expressions in attributes
         throw new Error('ICU expressions are not yet supported in attributes');
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n/i18n_parse.ts:243]: This is not directly related to this PR, but while we are changing this function we can also consider updating the logic related to ICUs in attributes. We can probably remove the inner `for` loop and just have a quick check (guarded by the `ngDevMode`) to verify that the value doesn't contain ICUs:
```
if (ngDevMode) {
  // Check if attribute value contains an ICU and throw an error if that's the case.
  // ICUs in element attributes are not supported.
  if (ICU_REGEXP.test(value)) {
    throw new Error('ICU expressions are not supported in attributes');
  }
}
```

**Review by AndrewKushnir — COMMENTED** [review]
Just left a couple comments, please see below.

> Also something I was thinking about while addressing the feedback: would it make sense to rename the i18nAttribute instruction to something like i18nDynamicAttribute to be more explicit about its uses?

Yeah, that makes sense, let's have a followup PR with the rename (once this PR lands). The "dynamic" sounds good, I was also thinking if we could call it "bound" (i.e. `i18nBoundAttributes`), which might be more aligned with the compiler terminology (where we have `t.BoundAttribute`).

**Comment by petebacondarwin** [comment]
+1 for `i18nBoundAttributes`

File: packages/compiler/src/render3/view/template.ts [hunk 11]
```diff
@@ -606,8 +601,12 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
         stylingBuilder.registerStyleAttr(value);
       } else if (name === 'class') {
         stylingBuilder.registerClassAttr(value);
+      } else if (isBoundI18nAttribute(attr)) {
```
Comment (petebacondarwin) [inline packages/compiler/src/render3/view/template.ts:604]: Something for another PR... but our typings are a bit off in the compiler.
`Element.attributes` is typed as `TextAttribute[]`, but it appears that it can hold `BoundAttribute` objects too.
Looking through the code there are a bunch of casts that hide this. 
It would be good to tighten this up later.

**Review by petebacondarwin — COMMENTED** [review]
Nice bit of refactoring. I agree with @AndrewKushnir's comments and added a few more.
Thanks for jumping on this @crisbeto!

File: packages/core/src/render3/i18n/i18n_parse.ts (hunk 10 above)
Comment (crisbeto) [reply packages/core/src/render3/i18n/i18n_parse.ts]: I'm not sure that it's completely safe to remove this from production code since the value could come from a translation file that may have an ICU in one language, but not the other.

**Comment by crisbeto** [comment]
I've addressed the latest set of feedback.

File: packages/core/src/render3/i18n/i18n_parse.ts [hunk 12]
```diff
@@ -225,51 +223,31 @@ function i18nStartFirstCreatePassProcessTextNode(
...
-          }
-        }
+  if (tView.firstCreatePass && tView.data[index] === null) {
+    for (let i = 0; i < values.length; i += 2) {
+      const attrName = values[i];
+      const message = values[i + 1];
+
+      // Check if attribute value contains an ICU and throw an error if that's the case.
+      // ICUs in element attributes are not supported.
+      if (ICU_REGEXP.test(message)) {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n/i18n_parse.ts:240]: > I'm not sure that it's completely safe to remove this from production code since the value
> could come from a translation file that may have an ICU in one language, but not the other.

GitHub UI doesn't let me reply to this message, so I'm replying here.

That's a great point @crisbeto 👍

Let's may be add a quick comment here to describe that:

```
      // ...
      // Note: we intentionally avoid using `ngDevMode` here (thus retaining this check in production build),
      // since translations may have different content, so we need to verify there are no ICUs in translated text.
      if (ICU_REGEXP.test(message)) {
        throw new Error(`ICU expressions are not supported in attributes. Message: "${message}".`);
      }
```

Note: I've also updated the error message text to include the message itself, so that the error message is more helpful to users.

For consistency we should also update the error [here](https://github.com/angular/angular/blob/master/packages/core/src/render3/i18n/i18n_parse.ts#L105-L107) to remove `ngDevMode` and add a note that it's intentional. @crisbeto could you please help with that?

------

Just as an idea for the future: we should be able to refactor these errors more once #39188 lands and exclude message text from prod bundles, but keep the error code, so the prod bundle doesn't contain long strings and at the same time, provide useful info:

```
      if (ICU_REGEXP.test(message)) {
        throw new RuntimeError(
          RuntimeErrorCode.UNSUPPORTED_ICU_IN_ATTRS,
          ngDevMode ? `ICU expressions are not supported in attributes. Message: "${message}".` : message);
      }
```

**Review by AndrewKushnir — COMMENTED** [review]
Thanks for the updates @crisbeto 👍

It's really great to see how the `i18nAttributesFirstPass` function is now simplified!

**Review by AndrewKushnir — APPROVED** [review]
LGTM, thanks @crisbeto 👍

**Review by petebacondarwin — APPROVED** [review]
Nice!

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:339170018:BASE:339170027:1603767647214:980f6de4).

File: packages/compiler/src/render3/view/template.ts [hunk 13]
```diff
@@ -914,7 +915,7 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
     // Only add normal input/output binding instructions on explicit <ng-template> elements.
     if (template.tagName === NG_TEMPLATE_TAG_NAME) {
       const [i18nInputs, inputs] = partitionArray(template.inputs, hasI18nMeta);
```
Comment (AndrewKushnir) [inline packages/compiler/src/render3/view/template.ts:917]: FYI, tests in Google's codebase returned the following build failure:
```
packages/compiler/src/render3/view/template.ts:931:54 - error TS2345: Argument of type '(Node & { i18n?: Message | Node | undefined; })[]' is not assignable to parameter of type '(TextAttribute | BoundAttribute)[]'.
  Type 'Node & { i18n?: Message | Node | undefined; }' is not assignable to type 'TextAttribute | BoundAttribute'.
    Type 'Node & { i18n?: Message | Node | undefined; }' is not assignable to type 'BoundAttribute'.

931         this.templatePropertyBindings(templateIndex, inputs);
                                                         ~~~~~~
```
changing the code as follows resolved the issue:
```suggestion
      const [i18nInputs, inputs] = partitionArray<t.BoundAttribute, t.BoundAttribute>(template.inputs, hasI18nMeta);
```
@crisbeto could you please apply this update? Thank you.

Comment (crisbeto) [reply packages/compiler/src/render3/view/template.ts:917]: Done.

Comment (AndrewKushnir) [reply packages/compiler/src/render3/view/template.ts]: Thanks @crisbeto 👍

**Comment by AndrewKushnir** [comment]
[Presubmit #2](http://test/OCL:339255186:BASE:339255210:1603813033942:132ef902).

**Comment by alxhub** [comment]
Merging to RC for now - please open a separate patch PR if this needs to merge to the patch branch.
